reddit_service_websockets/socketserver.py

```diff
@@ -150,8 +184,52 @@ def __call__(self, environ, start_response):
         finally:
             self.metrics.counter("conn.lost").increment()
             self._send_message("disconnect", {"namespace": namespace})
+            self.connections.remove(websocket)
             dispatcher.kill()
 
+    def _authorized(self, environ, admin_auth):
+        auth_header = environ.get('HTTP_AUTHORIZATION', None)
+        if not auth_header:
+            return False
+        auth_scheme, sep, auth_token = auth_header.strip().partition(' ')
+        assert auth_scheme.lower() == 'basic'
+        return auth_token == admin_auth
+
+    def _quiesce(self, environ):
+        """Set service state to quiesced and shed existing connections."""
+        if not self._authorized(environ, self.admin_auth):
+            raise UnauthorizedError
+
+        # Delay shedding to allow service deregistration after quiescing
+        shed_delay_secs = 30
+        total_conns = len(self.connections)
+        conns = self.connections.copy()
+        if not self.quiesced:
+            # Shed shed_rate_per_sec connections every second
+            #   after service deregistration delay.
+            cur_iter_sec = 0
+            for remaining in xrange(total_conns, 0, -self.shed_rate_per_sec):
+                cur_iter_sec += 1
+                # Check if fewer than shed_rate_per_sec conns left
+                #   in set so there's no over-popping.
+                if remaining >= self.shed_rate_per_sec:
+                    num_conns = self.shed_rate_per_sec
+                else:
+                    num_conns = remaining
+                gevent.spawn_later(cur_iter_sec + shed_delay_secs,
+                                   self._shed_connections,
+                                   [conns.pop() for j in xrange(num_conns)])
+            self.quiesced = True
```

Comment:
oh good call, I need to move the flag flipping and check before I make a copy of the connections.